source/agora/common/Metadata.d

```diff
@@ -0,0 +1,137 @@
+/*******************************************************************************
+
+    Defines metadata which is stored to disk (e.g. peer info)
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.common.Metadata;
+
+import agora.common.Data;
+import agora.common.Set;
+
+import vibe.core.log;
+import vibe.data.json;
+
+import std.array;
+import std.file;
+import std.format;
+import std.meta;
+import std.stdio;
+import std.typecons;
+
+/*******************************************************************************
+
+    Contains the serialized set of peers a node has connected to.
+    Note: Cannot be abstract, vibe.d requirement.
+
+*******************************************************************************/
+
+public class Metadata
+{
+    /// The set of peers we previously established connection to
+    public Set!Address peers;
+
+    /// Deserialize the Metadata from a Json string
+    public static void fromJsonString (Metadata metadata, string json)
```

Comment:
I'm not soooooo found of that design. Mixing `class` and JSON object usually doesn't end well.

The way I tend to look at it is that classes have an "identity" as two bitwise-equal objects are different, but value types have no identity as they are defined by the sum of their content.
For JSON object you want value type, not something with an identity.

Reply:
Well I can't use structs if I want to be able to provide Metadata instance that doesn't use IO and is used during unittests. What do you propose instead?